### PR TITLE
verbose errors feature

### PR DESCRIPTION
**Pull request by killercup**
This adds a new "verbose-errors" feature flag to async-std that enables
wrapping certain errors in structures with more context. As an example,
we use it in `fs::File::{open,create}` to add the given path to the
error message (something that is lacking in std to annoyance of many).

### CODE REVIEW

**Comment by killercup** [comment]
We discussed this a while back, and I wanted to provide something small that can be a basis for discussion. The main thing here is to provide an internal mechanism to easily add optional contexts. I hope–but have not proven–that the closure get compiled away in release modes without the feature flag set.

**Comment by stjepang** [comment]
This is great, thank you! :) I think we don't even need to introduce a new feature flag for this and can just provide verbose errors by default. Is there any real downside to providing this by default?

Wdyt, @yoshuawuyts ?

**Comment by killercup** [comment]
It costs a string allocation in the error case :shrug:

**Comment by stjepang** [comment]
Error cases don't need to be fast :)

Besides, we already have a string allocation for the `path`, a task allocation for `spawn`, a push into the queue, and so on... so the cost of this additional error context is practically zero.

File: src/utils.rs [hunk 1]
```diff
@@ -52,6 +52,20 @@ pub fn random(n: u32) -> u32 {
     })
 }
 
+/// Add additional context to errors
+///
+/// *Note for implementors:* The given closure must only be executed when
+/// `verbose-errors` feature is enabled for this crate!
+pub(crate) trait VerboseErrorExt: Sized {
```
Comment (yoshuawuyts) [inline src/utils.rs]: I think we could follow `anyhow` / `fehler`'s lead here and call it `Context`, with a method named `context`.

File: Cargo.toml [hunk 2]
```diff
@@ -48,6 +48,7 @@ std = [
   "pin-utils",
   "slab",
 ]
+verbose-errors = []
```
Comment (yoshuawuyts) [inline Cargo.toml]: Agree we don't need a feature for this (:

File: src/utils.rs (hunk 1 above)
Comment (killercup) [reply src/utils.rs]: Agreed now that it's no longer the name of the feature flag :)

File: tests/verbose_errors.rs [hunk 3]
```diff
@@ -0,0 +1,20 @@
+#[cfg(feature = "verbose-errors")]
```
Comment (otavio) [inline tests/verbose_errors.rs]: The flag was removed, wasn't it?

Comment (killercup) [reply tests/verbose_errors.rs]: A sorry, forgot this one. Fixed.

Comment (killercup) [reply tests/verbose_errors.rs]: A sorry, forgot this one. Fixed.

**Comment by killercup** [comment]
🎉 let's maybe open an issue to track more places we can add this

**Pull request merged**